rust/noosphere/src/ffi/fs.rs

```diff
@@ -0,0 +1,233 @@
+use anyhow::anyhow;
+use noosphere_core::data::Did;
+use noosphere_fs::{SphereFile, SphereFs};
+use safer_ffi::prelude::*;
+use std::{pin::Pin, str::FromStr};
+use subtext::{Peer, Slashlink};
+use tokio::io::{AsyncRead, AsyncReadExt};
+
+use crate::{
+    ffi::{NsHeaders, NsNoosphereContext},
+    platform::{PlatformKeyMaterial, PlatformStore},
+};
+
+ReprC! {
+    #[ReprC::opaque]
+    pub struct NsSphereFs {
+        inner: SphereFs<PlatformStore, PlatformKeyMaterial>
+    }
+}
+
+impl NsSphereFs {
+    pub fn inner(&self) -> &SphereFs<PlatformStore, PlatformKeyMaterial> {
```

Comment:
I wonder: could these be traits?
`trait NSFFI<T> { fn inner(&self) -> &T; fn inner_mut(&mut self) -> &mut T; }`

Reply:
I don't think so. `safer-ffi` doesn't allow generic bounds in FFI calls, and also loses meaning at the FFI boundary anyway.

Maybe at some point in the future we will support multiple backends on a given platform, but at this point the use case isn't clear. So, I'm inclined to lean into the simplicity of concrete types at this boundary.